mono/metadata/boehm-gc.c

```diff
@@ -600,7 +600,7 @@ push_handle_stack (HandleStack* stack)
 
 	while (cur) {
 		if (cur->size > 0)
-			GC_push_all (cur->elems, (char*)(cur->elems + cur->size) + 1);
+			GC_push_all ((gpointer)&cur->elems[0], (char*)(cur->elems + cur->size) + 1);
```

Comment:
This is the wrong fix. Change the function to take void* instead of char*. As well, this fix is not valid C++. As well, subjectively, once you fix gc_push_all, I'd pass &cur->elems, which is a strange type (pointer to sized-array without decay-to-pointer), but reads well, and works.

Reply:
Like this, sufficient but not the entire thing:

commit 0bdcfb661c956a79f2609fda57e8b99efb4410bb
Author: Jay Krell <jaykrell@microsoft.com>
Date:   Sun Jan 28 00:51:25 2018 -0800

    Fixing more code to be C++.

diff --git a/libgc/include/private/gc_priv.h b/libgc/include/private/gc_priv.h
index 53afd7828d1..de3c9169f47 100644
--- a/libgc/include/private/gc_priv.h
+++ b/libgc/include/private/gc_priv.h
@@ -1315,7 +1315,7 @@ void GC_initiate_gc GC_PROTO((void));
                                /* If the mark state is invalid, this   */
                                /* becomes full colleection.  Otherwise */
                                /* it's partial.                        */
-void GC_push_all GC_PROTO((ptr_t bottom, ptr_t top));
+void GC_push_all (void* bottom, void* top);
                                /* Push everything in a range           */
                                /* onto mark stack.                     */
 void GC_push_selected GC_PROTO(( \
diff --git a/libgc/mark.c b/libgc/mark.c
index e870342e412..75ec282160d 100644
--- a/libgc/mark.c
+++ b/libgc/mark.c
@@ -1191,11 +1191,11 @@ void GC_mark_init()
  * Should only be used if there is no possibility of mark stack
  * overflow.
  */
-void GC_push_all(bottom, top)
-ptr_t bottom;
-ptr_t top;
+void GC_push_all(void* void_bottom, void* void_top)
 {
-    register word length;
+    ptr_t bottom = (ptr_t)void_bottom;
+    ptr_t top = (ptr_t)void_top;
+    word length;
     
     bottom = (ptr_t)(((word) bottom + ALIGNMENT-1) & ~(ALIGNMENT-1));
     top = (ptr_t)(((word) top) & ~(ALIGNMENT-1));

Reply:
It doesn't make sense to change boehm - for one Mono is able to compile with an external Boehm GC (this is, for example, what the Unity fork does).

Reply:
Eek I thought they used the one in-tree. Alas ideally if code wasn't forked all over the place, void* would be the thing.